### ISSUE TITLE
0.13.0 breaks react-hot-loader

### ISSUE THREAD

**Issue by bobbyrenwick** [FormidableLabs/radium#219]
I'm using `React.createClass`, `Radium` and `react-hot-loader@1.2.7` like follows:

```es6
var React = require("react");
var Radium = require("radium");

var SearchInput = React.createClass({

    getStyles: function() {
        const styles = {
            base: {
                backgroundColor: "red",
            },
        };

        return [
            styles.base,
        ];
    },

    render: function() {
        return (
            <div style={this.getStyles()}>
                Hello Radium
            </div>
        );
    }

});

module.exports = Radium(SearchInput);
```

When I updated to `v0.13.0`, hot updates to components using Radium stopped working. It was working with `v0.12.1` before. This is a pretty big as all my styles are done with Radium and this now stops me from seeing the effect of my code changes.

**Comment by ianobermiller** [comment]
Hmm, react-hot-loader seems to work fine when using decorators and ES2015 class syntax. Might be something specific to createClass?

**Comment by epsitec** [comment]
I was not aware of the fact that it was Radium's implementation broke hot reloader. But I can confirm that our hot reloading is no longer working either.

**Comment by ianobermiller** [comment]
@epsitec, are you using createClass or classes?

**Comment by alexlande** [comment]
As another anecdote-- hot-loader is working fine for me on a current project using decorator syntax with ES2015 classes.

**Comment by reinman** [comment]
Hotloader + ES5 + Radium 0.13.0 is a no-go for me right now... downgraded to 0.12.2 and things are peachy again.

**Comment by epsitec** [comment]
We are using `createClass` but the issue might not be related to Radium. I still have to nail it down, as we have another layer also wrapping the React classes before passing them down to Radium.

**Comment by bobbyrenwick** [comment]
could it be when combining Radium and PureRenderMixin? @epsitec is that what your extra wrapping layer does?

**Comment by epsitec** [comment]
No PureRenderMixin in our case. We are using the code found [here](https://gist.github.com/epsitec/d28e8f9fe2c6533b1111)

**Comment by bobbyrenwick** [comment]
@ianobermiller I'm using my master branch and this no longer seems to be an issue.

**Comment by bobbyrenwick** [comment]
Hmm, maybe I was a little hasty there, updates to styles works, but adding new functions and calling them in render doesn't work.

**Comment by bobbyrenwick** [comment]
OK, so that seems to be the issue, any change in a function other than render doesn't get picked up. @epsitec, @reinman are you using functions to retrieve your styles?

As soon as I put styles in object literals straight into the JSX, the hot loading works again

**Comment by ianobermiller** [comment]
This might be a general issue with Higher Order Components and hot
reloading then. @gaearon, have you seen any issues in this realm?

**Comment by gaearon** [comment]
Components wrapped in HOCs will lose their state on hot reload with the current implementation. I plan to solve this eventually with a Babel plugin.

However hot reloading should still generally *work* with HOCs—it's just that the state is lost. If it does not *work*, I'd appreciate somebody forking https://github.com/gaearon/react-hot-boilerplate and adding Radium there so I can reproduce the issue and take a look.

**Comment by bobbyrenwick** [comment]
@gaearon great talk at ReactEurope by the way. It was easily the most exciting thing I've seen since I started playing around with React!

I've forked your hot boilerplate, added Radium and converted `App` to use `React.createClass` here https://github.com/bobbyrenwick/react-hot-boilerplate/tree/test-radium-hoc. If you change the `color` from `'red'` to `'blue'` you'll notice no change. But if you remove the Radium wrapper it will work.

**Comment by ianobermiller** [comment]
@bobbyrenwick :trophy:

**Comment by gaearon** [comment]
Hey, thanks so much. I'll look into this soon.

**Comment by gaearon** [comment]
I looked into this.

React Hot Loader does not support replacing inherited methods. It only works with “top-level” prototype methods. The use case was so rare (people rarely inherit components and then change inherited methods) so I never implemented that.

It is also very hard to implement correctly. Inherited components need to maintain correct `super` method references, so now we need to proxy each prototype in chain and remember which proxied methods correspond to which prototype. Much more difficult than what RHL normally does.

So, the issue is that, instead of wrapping the inner component as a higher-order component, Radium actually inherits from it and thus hides its methods. Some possible ways to fix it:

* Make Radium a real HOC
* Change Radium to manually copy user-defined methods to the exported prototype
* Export the unwrapped class manually in every file using Radium

I'm not sure there's anything I can do in RHL to help fix this.

**Comment by gaearon** [comment]
The third one is an ugly (but totally usable) work around for now: you can replace

```js
module.exports = Radium(App);
```

with

```js
module.exports = Radium(App);
module.exports.App = App;
```

and be happy for now.

Even moreso, if you're not using ES6, you can actually put the original component in statics of the `Radium`-generated wrapper, and it will “just work” for ES5 code.

**Comment by ianobermiller** [comment]
@gaearon thanks greatly for looking into this!

Option 2 ("Change Radium to manually copy user-defined methods to the exported prototype") isn't so bad, and we could only do it dev mode in the presence of hot loading if we wanted.

Not sure if Radium can be a true HOC (rendering the inner) since it depends on mucking with the rendered elements.

**Comment by gaearon** [comment]
I think it should also work if instead of subclassing you do something like

```js
let { componentWillReceiveProps } = ComposedComponent.prototype;
ComposedComponent.prototype.componentWillReceiveProps = function () {
  if (componentWillReceiveProps) {
    componentWillReceiveProps.apply(this, arguments);
  }
  // ...
}

// other methods same way
```

**Comment by bobbyrenwick** [comment]
This final option looks pretty clean to me @gaearon! @ianobermiller that way we can remove the IE < 10 `__proto__` hacks too.

**Comment by ianobermiller** [comment]
Funny enough that is basically what the old createClass code did. I think hot loading is important enough that I'd be ok with this implementation. Should be slightly faster as well. 

One downside is that Radium would mutate the class you pass into it instead of returning a new class. This may not be a real problem, but something to keep in mind. Definitely a breaking change, since `Radium(MyComponent)` now has a different meaning.

**Comment by ianobermiller** [comment]
Fixed in #255
